HW2/P3/mandelbrot.pyx

```diff
@@ -32,17 +39,70 @@ cpdef mandelbrot(np.complex64_t [:, :] in_coords,
     assert in_coords.shape[1] == out_counts.shape[1],  "Input and output arrays must be the same size"
 
     with nogil:
-        for i in range(in_coords.shape[0]):
-            for j in range(in_coords.shape[1]):
-                c = in_coords[i, j]
-                z = 0
+        # Set the limit of the mandelbrot function
+        max_mag = AVX.float_to_float8(4.0)
+
+        # Set one floats to add to iterations for coordinates
+        one_bits = AVX.float_to_float8(1.0)
+
+        for i in prange(in_coords.shape[0], num_threads=4, schedule='static', chunksize=1):
+            for j in range(0, in_coords.shape[1], 8):
+                # Grab imaginary coordinates 
+                coords_imag = AVX.make_float8(in_coords[i,j+7].imag,
+                                              in_coords[i,j+6].imag,
+                                              in_coords[i,j+5].imag,
+                                              in_coords[i,j+4].imag,
+                                              in_coords[i,j+3].imag,
+                                              in_coords[i,j+2].imag,
+                                              in_coords[i,j+1].imag,
+                                              in_coords[i,j].imag)
+                
+                # Grab real coordinates
+                coords_real = AVX.make_float8(in_coords[i,j+7].real,
+                                              in_coords[i,j+6].real,
+                                              in_coords[i,j+5].real,
+                                              in_coords[i,j+4].real,
+                                              in_coords[i,j+3].real,
+                                              in_coords[i,j+2].real,
+                                              in_coords[i,j+1].real,
+                                              in_coords[i,j].real)
+                
```

Comment:
Code really well commented and easy to follow. Here I chose to convert in batch the whole array in real and imaginary part at the beginning and then to retrieve value from this array, taking more memory but then less instructions to perform I think.